File: shared/ts/src/auth/middleware.ts
import { toErrorResponse } from '../errors';
import { buildAuditEvent, recordAuditEvent } from './audit';
import type { PolicyEngine } from './policy';

export interface AuthActor {
  subject: string;
  roles: string[];
}

export type ActorExtractor = (request: { headers: Record<string, unknown> }) => AuthActor;

export const headerExtractor: ActorExtractor = (request) => {
  const subject = String(request.headers['x-actor-subject'] ?? '');
  const rawRoles = String(request.headers['x-actor-roles'] ?? '');
  const roles = rawRoles
    .split(',')
    .map((role) => role.trim())
    .filter(Boolean);
  return { subject, roles };
};

export interface RequestLike {
  method: string;
  url: string;
  headers: Record<string, unknown>;
  log?: { warn?: (msg: unknown) => void };
}

export interface ReplyLike {
  status: (code: number) => ReplySender;
}

export interface ReplySender {
  send: (payload: unknown) => void;
}

export function createAuthMiddleware(
  engine: PolicyEngine,
  extractor: ActorExtractor = headerExtractor,
) {
  return async function authorize(
    request: RequestLike,
    reply: ReplyLike,
  ) {
    const actor = extractor(request);
    const action = `${request.method.toUpperCase()}:${request.url}`;
    const allowed = engine.isAllowed(action, actor.roles);
    recordAuditEvent(buildAuditEvent(action, actor, allowed));

    if (!allowed) {
      request.log?.warn?.({ actor, message: 'authorization denied' });
      const response = toErrorResponse(
        new Error('access denied'),
      );
      response.code = 'UNAUTHORIZED';
      response.actor = actor;
      response.request_id = String(request.headers['x-request-id'] ?? '');
      const sender = reply.status(403);
      sender.send(response);
      return reply;
    }
    return undefined;
  };
}

File: shared/go/auth/middleware.go
package auth

import (
	"context"
	"net/http"
	"strings"

	"github.com/ai-aas/shared-go/errors"
)

type contextKey string

const actorContextKey contextKey = "shared.auth.actor"

// Actor represents the authenticated subject attached to a request.
type Actor struct {
	Subject string
	Roles   []string
}

// ActorFromContext extracts the actor from request context.
func ActorFromContext(ctx context.Context) (Actor, bool) {
	actor, ok := ctx.Value(actorContextKey).(Actor)
	return actor, ok
}

// Extractor derives an actor from an HTTP request.
type Extractor func(*http.Request) Actor

// HeaderExtractor reads actor information from X-Actor-* headers.
func HeaderExtractor(r *http.Request) Actor {
	subject := r.Header.Get("X-Actor-Subject")
	rawRoles := r.Header.Get("X-Actor-Roles")
	var roles []string
	for _, role := range strings.Split(rawRoles, ",") {
		role = strings.TrimSpace(role)
		if role != "" {
			roles = append(roles, role)
		}
	}
	return Actor{
		Subject: subject,
		Roles:   roles,
	}
}

// Middleware enforces authorization using the supplied engine.
func Middleware(engine *Engine, extractor Extractor) func(http.Handler) http.Handler {
	if extractor == nil {
		extractor = HeaderExtractor
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			actor := extractor(r)
			action := r.Method + ":" + r.URL.Path
			allowed := engine.Allowed(action, actor.Roles)
			recordAudit(NewAuditEvent(action, actor, allowed))

			if !allowed {
				resp := errors.New("UNAUTHORIZED", "access denied",
					errors.WithActor(&errors.Actor{
						Subject: actor.Subject,
						Roles:   actor.Roles,
					}),
					errors.WithRequestID(r.Header.Get("X-Request-ID")),
				)
				data, _ := errors.Marshal(resp)
				w.Header().Set("Content-Type", "application/json")
				w.WriteHeader(http.StatusForbidden)
				_, _ = w.Write(data)
				return
			}

			ctx := context.WithValue(r.Context(), actorContextKey, actor)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}
